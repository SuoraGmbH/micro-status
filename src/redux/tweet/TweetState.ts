import { Tweet } from '../../domain/Tweet';
import { ApplicationState, tweetState } from '../ApplicationState';

export interface TweetState {
  byId: { [key: number]: Tweet };
  all: number[];
}

export const getTweetById = (state: ApplicationState, tweetId: number): Tweet =>
  tweetState(state).byId[tweetId];

export const getAllTweets = (state: ApplicationState): Tweet[] =>
  tweetState(state).all.map(tweetId => getTweetById(state, tweetId));
