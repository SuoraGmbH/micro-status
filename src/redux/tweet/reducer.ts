import { Action } from '../Action';
import { TweetState } from './TweetState';
const initialState: TweetState = {
  byId: {},
  all: [],
};
export const tweetReducer = (
  state: TweetState = initialState,
  action: Action,
) => {
  switch (action.type) {
    case 'FETCH_TWEETS_FULFILLED':
      const byId = { ...state.byId };
      const all = [...state.all];

      action.payload.forEach(tweet => {
        byId[tweet.id] = tweet;
        if (!all.includes(tweet.id)) {
          all.push(tweet.id);
        }
      });

      return {
        ...state,
        all,
        byId,
      };
    case 'ADD_TWEET_FULFILLED':
      const tweet = action.payload;
      return {
        ...state,
        byId: { ...state.byId, [tweet.id]: tweet },
        all: [...state.all, tweet.id],
      };
  }
  return state;
};
