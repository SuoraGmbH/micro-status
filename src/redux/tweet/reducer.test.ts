import { tweetReducer } from './reducer';
import { TweetState } from './TweetState';

describe('tweetReducer', () => {
  let initialState: TweetState;
  const tweetFromTobmaster = {
    id: 1,
    date: '2018-09-23T17:52:21+0200',
    message: 'Hello Tests',
    userId: 'tobmaster',
  };

  const tweetFromMuhdiekuh = {
    id: 2,
    date: '2018-09-24T17:52:21+0200',
    message: 'Hello other test!',
    userId: 'muhdiekuh',
  };

  const tweetFromNiklasWithIdFromMuhdiekuh = {
    id: 2,
    date: '2018-09-25T17:52:21+0200',
    message: 'Hello angular!',
    userId: 'niklas_wortmann',
  };

  beforeAll(() => {
    initialState = tweetReducer(undefined, { type: '@@INIT' });
  });

  it('should have a valid initial state', () => {
    expect(initialState).toMatchSnapshot();
  });

  it('should have tweets after FETCH_TWEETS_FULFILLED', () => {
    const state = tweetReducer(initialState, {
      type: 'FETCH_TWEETS_FULFILLED',
      payload: [tweetFromTobmaster],
    });

    expect(state.byId).toEqual({ 1: tweetFromTobmaster });
    expect(state.all).toEqual([1]);
  });

  it('should support multiple FETCH_TWEETS_FULFILLED actions and contain all the containing tweets', () => {
    let state = tweetReducer(initialState, {
      type: 'FETCH_TWEETS_FULFILLED',
      payload: [tweetFromTobmaster],
    });

    state = tweetReducer(state, {
      type: 'FETCH_TWEETS_FULFILLED',
      payload: [tweetFromMuhdiekuh],
    });

    expect(state.byId).toEqual({
      1: tweetFromTobmaster,
      2: tweetFromMuhdiekuh,
    });
    expect(state.all).toEqual([1, 2]);
  });

  it('should overwrite a tweet with new data', () => {
    let state = tweetReducer(initialState, {
      type: 'FETCH_TWEETS_FULFILLED',
      payload: [tweetFromMuhdiekuh],
    });

    state = tweetReducer(state, {
      type: 'FETCH_TWEETS_FULFILLED',
      payload: [tweetFromNiklasWithIdFromMuhdiekuh],
    });

    expect(state.byId).toEqual({
      2: tweetFromNiklasWithIdFromMuhdiekuh,
    });

    expect(state.all).toEqual([2]);
  });
});
