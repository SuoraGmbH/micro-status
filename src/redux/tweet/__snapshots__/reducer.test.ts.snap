// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tweetReducer should have a valid initial state 1`] = `
Object {
  "all": Array [],
  "byId": Object {},
}
`;
